src/django_idom/websocket_consumer.py

```diff
@@ -15,6 +18,12 @@
 _logger = logging.getLogger(__name__)
 
 
+@dataclass
+class WebSocketConnection:
+    scope: dict
+    disconnect: Callable
```

Comment:
Can you make this a bit more explicit? Is this supposed to be `Callable[[], None]`?

Reply:
Disconnect is a function we have within Django IDOM's consumer. So it'll never be none.

Reply:
What `Callable[[], None]` indicates is that the intended usage is `ws_conn.disconnect()` where the return value of `disconnect()` is `None`.

Given that it comes directly from Django IDOM's consumer, the signature would be `Callable[[int], Awaitable[None]]` with the intended use `await ws_conn.disconnect(code)`. This is fine, but we should probably have a sync version of this function too called `disconnect_soon`. This would be a a simple wrapper around `disconnect` that uses `asyncio.ensure_future`.

The interface here should then be:

```python
@dataclass
class WebSocketConnection:
    scope: dict
    disconnect: Callable[[int], typing.Awaitable[None]]
    disconnect_soon: Callable[[int], asyncio.Future[None]]
```

The implementation for `disconnect_soon` ought to look something like:

```python
def disconnect_soon(code: int) -> asyncio.Future[None]:
    return asyncio.ensure_future(disconnect(code))
```

Reply:
I don't think a sync helper will be needed since we'll always prefer calling async within our consumer.

Reply:
You won't be able to call the async function from within a component directly though since that's a sync function. You'd need to do so in an event handler of component effect hook.

Reply:
The django channels recommended way of using async within a sync context is using their conversion helper. We could simply recommend using it to minimize function wrappers.

```python 
from channels.db import database_sync_to_async, sync_to_async
```

---

With that in mind, has there been considerations for async components within IDOM core? It's theoretically possible for someone to have a long polling component action. Would be best if that doesn't lock up IDOM and Django Channels.

For example within Conreq I allow most things to be either via inspect.

```python
from inspect import iscoroutinefunction

async def sign_up(request):
    view = app.config.sign_up_view
    if iscoroutinefunction(view):
        return await view(request)
    return view(request)
```

Reply:
So in this case I don't think those helpers from Django apply here since what we'd want is one that converts an async function to a sync one which is what `asyncio.ensure_future` is meant to do.

As for async components, in the past they actually were async. However there are two reasons they aren't now:

1. **Rendering the layout is a serial task** - technically you could render sibling components in parallel, but in 99% of cases that has no benefit at the cost of making things more complicated for the author to understand.
2. **You can use async functions in the [`use_effect`](https://idom-docs.herokuapp.com/docs/life-cycle-hooks.html#async-effects) hook** - an action like triggering the websocket to disconnect is a perfect example of a "side effect" that belongs withing the `use_effect` hook.

Reply:
Apologies for the late reply.

Django ASGI-REF has the inverse of the previously mentioned function. For example: [async_to_sync](https://docs.djangoproject.com/en/3.2/topics/async/#asgiref.sync.sync_to_async). I don't think the developer manually disconnecting the websocket is common enough scenario to justify overcomplicating our function helpers. Frankly the only scenario I can think of it being utilizing is force disconnecting the WS after failed user authentication.

---

The discussion about async within IDOM should probably be moved to another thread but I'll quickly post my thoughts.

I agree that in most cases there isn't much benefit to async components, and they add complexity as you mentioned. That's pretty much the same conclusion that the Django team reached with async views. However, I think long term supporting both is ideal in order to prevent blocking/polling tasks from stalling IDOM.

The solution the Django team took was to the developer to decide whether to use sync or async views. To my knowledge, sync views are run in threads, and async views are handled through an event loop.